core/api/make.cpp

```diff
@@ -310,17 +320,98 @@ namespace fc::api {
           }
           return result;
         }},
+        .StateListMessages = {[=](auto &match, auto &tipset_key, auto to_height)
+                                  -> outcome::result<std::vector<CID>> {
+          OUTCOME_TRY(context, tipsetContext(tipset_key));
+
+          // TODO(artyom-yurin): Make sure at least one of 'to' or 'from' is defined
+
+          auto matchFunc = [&](const UnsignedMessage &message) -> bool {
+            if (match.to != message.to) {
+              return false;
+            }
+
+            if (match.from != message.from) {
+              return false;
+            }
+
+            return true;
+          };
+
+          std::vector<CID> result;
+
+          while (static_cast<int64_t>(context.tipset.height) >= to_height) {
+            // TODO(turuslan): FIL-146 randomness from tipset
+            std::shared_ptr<RandomnessProvider> randomness;
```

Comment:
You dont need `Env` here. Just keep set of visited message cids.

Reply:
Removed and added `std::set` for collect unique CIDs